src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+    // in a blockchain should have this metadata.
+    if (operation.transactionNumber === undefined ||
+        operation.operationIndex === undefined ||
+        operation.batchFileHash === undefined) {
+      return null;
+    }
+
+    // opInfo is operation with derivable properties projected out
+    const opInfo: OperationInfo = {
+      transactionNumber: operation.transactionNumber,
+      operationIndex: operation.operationIndex,
+      batchFileHash: operation.batchFileHash,
+      type: operation.type
+    };
+
+    // If this is a duplicate of an earlier operation, we can
+    // ignore this operation. Note that we might have a previous
+    // operation with the same hash, but that previous operation
+    // need not be earlier in timestamp order - hence the check
+    // with lesser().
+    const prevOperation = this.opHashToInfo.get(opHash);
+    if (prevOperation !== undefined && lesser(prevOperation, opInfo)) {
+      return null;
+    }
+    // Update our mapping of operation hash to operation info overwriting
+    // previous info if it exists
+    this.opHashToInfo.set(opHash, opInfo);
+
+    // For operations that have a previous version, we need additional
+    // bookkeeping
+    if (operation.previousOperationHash) {
+      this.applyOpWithPrev(opHash, opInfo, operation.previousOperationHash);
+    }
+
+    return opHash;
+  }
+
+  public constructor (
+    private readonly cas: Cas,
+    private readonly didDocUpdate: DidDocumentUpdater,
+    private readonly didDocCreate: DidDocumentCreator) {
+
+  }
+
+  /**
+   * Rollback the state of the DidCache by removing all operations
+   * with transactionNumber greater than the provided parameter value.
+   * The intended use case for this method is to handle rollbacks
+   * in the blockchain.
+   *
+   * The current implementation is inefficient: It simply scans the two
+   * hashmaps storing the core Did state and removes all entries with
+   * a greater transaction number.  In future, the implementation should be optimized
+   * for the common case by keeping a sliding window of recent operations.
+   */
+  public rollback (transactionNumber: number) {
+
+    // Iterate over all nextVersion entries and remove all versions
+    // with "next" operation with transactionNumber greater than the provided
+    // parameter.
+    this.nextVersion.forEach((opHash, version, map) => {
+      const opInfo = this.opHashToInfo.get(opHash) as OperationInfo;
+      if (opInfo.transactionNumber > transactionNumber) {
+        map.delete(version);
+      }
+    });
+
+    // Iterate over all operations and remove those with with
+    // transactionNumber greater than the provided parameter.
+    this.opHashToInfo.forEach((opInfo, opHash, map) => {
+      if (opInfo.transactionNumber > transactionNumber) {
+        map.delete(opHash);
+      }
+    });
+  }
+
+  /**
+   * Resolve a did.
+   */
+  public async resolve (did: VersionId): Promise<DidDocument | null> {
```

Comment:
@csuwildcat made a point recently that DID is no should be hash of the initial DID Document, not hash of operation, so the type should just be `string`. So the code probably needs another look up map of DID -> createOperationHash. There is an [issue](https://github.com/decentralized-identity/sidetree-core/issues/38) created to modify code already written, but we should not commit more code that violates this requirement.

Reply:
This has a dependency on how we fix an earlier issue (copied here):

"Even if apply() does not return anything, the did cache needs to match previousOperationHash in an update operation with a previous operation that has this hash. This means that the cache should either (1) have the ability to compute hashes of operations or (2) the caller (e.g., observer) needs to provide the operation hash as part of an Operation.

If we go with (1) we need to provide the blockNumber in WriteOperation; if we go with (2) we need to provide operationHash in WriteOperation. I don't have a strong preference between (1) and (2)"

If we go with (2) all hashes are provided by observer (or rooter may be) and there is not much change to DidCache code. If we go with (1) what you say is correct. Holding off on this until we resolve (1) vs (2) offline.

Reply:
Created an issue to track this.